src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,296 @@
+// SPDX-FileCopyrightText: © 2020-2021 Lev Livnev <lev@liv.nev.org.uk>
+// SPDX-FileCopyrightText: © 2022 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.6.12;
+
+import {GemAbstract} from "dss-interfaces/ERC/GemAbstract.sol";
+import {DaiAbstract} from "dss-interfaces/dss/DaiAbstract.sol";
+import {PsmAbstract} from "dss-interfaces/dss/PsmAbstract.sol";
+import {GemJoinAbstract} from "dss-interfaces/dss/GemJoinAbstract.sol";
+
+/**
+ * @author Lev Livnev <lev@liv.nev.org.uk>
+ * @author Nazar Duchak <nazar@clio.finance>
+ * @title An Input Conduit for real-world assets (RWA).
+ * @dev This contract differs from the original [RwaInputConduit](https://github.com/makerdao/MIP21-RWA-Example/blob/fce06885ff89d10bf630710d4f6089c5bba94b4d/src/RwaConduit.sol#L20-L39):
+ *  - The caller of `push()` is not required to hold MakerDAO governance tokens.
+ *  - The `push()` method is permissioned.
+ *  - `push()` permissions are managed by `mate()`/`hate()` methods.
+ *  - Require PSM address in constructor
+ *  - The `push()` method swaps GEM to DAI using PSM
+ *  - THe `push()` method with `amount` argument swaps specified amount of GEM to DAI using PSM
+ *  - The `quit` method allows moving outstanding GEM balance to `quitTo`. It can be called only by the admin.
+ *  - The `quit` method with `amount` argument allows moving specified amount of GEM balance to `quitTo`.
+ *  - The `file` method allows updating `quitTo`, `to` addresses. It can be called only by the admin.
+ */
+contract RwaInputConduit3 {
+    /// @notice PSM GEM token contract address
+    GemAbstract public immutable gem;
+    /// @notice PSM contract address
+    PsmAbstract public immutable psm;
+    /// @dev DAI/GEM decimal difference.
+    uint256 private immutable to18ConvertionFactor;
+
+    /// @dev This is declared here so the storage layout lines up with RwaInputConduit.
+    address private __unused_gov;
+    /// @notice Dai token contract address
+    DaiAbstract public dai;
```

Comment:
Honestly is pretty weird `dai` is not immutable but `gem` is.
I think we should have made all the non fileable variables immutable since we started to use solc 0.6.x. Not sure why it didn't happen already in version 2.
I'd say we change it now to avoid carrying this thing over.

Reply:
`dai` is not immutable so we can keep the storage layout compatible with the other conduits.

Reply:
Yeah I know @godsflaw requested to try to maintain the storage slots.
I'm more from the opinion we should evolve and keep the code cleaner and more consistent within itself.
More in these cases where they are not contracts massively used but all the contrary.

Reply:
Personally I agree with you.

Storage layout should not be considered part of a contract's public API, it's an implementation detail.
It only really matters for things like `delegatecall` and the old `hevm.store()` from Dapp Tools that luckily can be replaced by Foundry's `stdstore` now.

Reply:
It might also be relevant for certain UIs or databases that grab raw data. However in this case it feels code clearness is more important as I seriously doubt we have tools analysing or grabbing MIP21 data.
Anyway nothing that I feel too strong either.

Reply:
Going back to this item and in concordance with what I asked for the Output one:
- I'd make `dai` immutable
- Remove `__unused_gov`
- Remove the `file`able option of `to` as now that we have the extended version of `yank` I don't really see the need to update it and then make it immutable as well.

Reply:
@godsflaw ^^

Reply:
This would also convert the `wards` as storage 0 which is a pretty common standard.

Reply:
> Remove the `file`able option of to as now that we have the extended version of yank I don't really see the need to update it and then make it immutable as well.

The `file`able to is not meant for one-off mistakes solved by `yank`. It shall be used if the `RwaUrn` address for a specific deal needs to change for whatever reason. That's a permanent-ish change, so it should probably remain here.

Reply:
Ok if this is something we are seeing as permanent change because the `urn` could be upgraded then sounds good.
I'd still suggest to move it after the mappings (as converting `dai` to immutable and removing `__unused_gov`).

Reply:
We are at odds with our recommendations here.  The storage layout preservation is primarily so integrations that read the storage layout, don't start ingesting bad data and corrupt their local databases.  This isn't an abstract consideration, this happened on multiple occasions with the VulcanizeDB folks.  If possible, especially where we know the contract is likely integrated in this way, we should try to preserve storage layout.

That being said, it is probably less likely that these contracts are integrated (beyond the public getters and events) in a way that will cause DB corruption.  One could make they argument that they aren't really upgrades from an integrations perspective, as the old contracts are not deprecated.  I'm happy to yield on this point in this particular case, mostly because each urn and conduit gets its own deployment with bespoke business logic.  But let's keep it in mind that, for more core contracts, not preserving storage layout can cause quite the headache for integrations.

Reply:
Yeah I think depends on each case, in this one IMO seems better code cleanness than worrying about storage slots standard. So having this response from Mooney, I think from PE side we can agree it is fine to follow these changes.